stan/math/rev/core/var.hpp

```diff
@@ -63,185 +74,65 @@ class var {
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(short x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(int x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(long x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  var(unsigned char x)  // NOLINT(runtime/explicit)
-      : vi_(new vari(static_cast<double>(x), false)) {}
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  // NOLINTNEXTLINE
-  var(unsigned short x) : vi_(new vari(static_cast<double>(x), false)) {}
-
-  /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
-   */
-  // NOLINTNEXTLINE
-  var(unsigned int x)
-      : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
+  var_value() : vi_(static_cast<vari_type*>(nullptr)) {}
 
   /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
+   * Construct a variable from the specified arithmetic type argument
+   * by constructing a new `vari_value<value_type>`. For integral types the
+   * `vari_value<value_type>` will hold doubles. This constructor is only valid
+   * when `S` is convertible to this `vari_value`'s `value_type`.
+   * @tparam S A type that is convertible to `value_type`.
    * @param x Value of the variable.
    */
-  // NOLINTNEXTLINE
-  var(unsigned long x)
-      : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-#ifdef _WIN64
-  // these two ctors are for Win64 to enable 64-bit signed
-  // and unsigned integers, because long and unsigned long
-  // are still 32-bit
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
```

Comment:
According to https://en.cppreference.com/w/cpp/types/is_convertible

"If the imaginary function definition To test() { return std::declval<From>(); } is well-formed, (that is, either std::declval<From>() can be converted to To using implicit conversions"

So ```var_value(T x)``` is the same thing without the templates.

Reply:
This constructor is not explicit so it's used for assignment. Removing this template causes a failure like 


```
./stan/math/rev/fun/grad_inc_beta.hpp: In function ‘void 
stan::math::grad_inc_beta(stan::math::var&, stan::math::var&,
  const var&, const var&, const var&)’:
./stan/math/rev/fun/grad_inc_beta.hpp:42:13: error: 
  conversion from ‘int’ to ‘stan::math::var’ 
  {aka ‘stan::math::var_value<double>’} is ambiguous
   42 |   var dF1 = 0;
      |             ^
In file included from ./stan/math/rev/meta/is_var.hpp:4,
./stan/math/rev/core/var.hpp:96:3: note:
   candidate: ‘stan::math::var_value<T, typename 
     std::enable_if<std::is_floating_point<typename std::decay<_Tp>::type>::value, 
       void>::type>::var_value(stan::math::vari_value<T>*) 
       [with T = double; 
        typename std::enable_if<std::is_floating_point<typename std::decay<_Tp>::type>::value, void>::type = void]’
   96 |   var_value(vari_value<T>* vi)  // NOLINT
      |   ^~~~~~~~~
./stan/math/rev/core/var.hpp:87:3: note:
 candidate:  ‘stan::math::var_value<T, typename 
    std::enable_if<std::is_floating_point<typename std::decay<_Tp>::type>::value,
    void>::type>::var_value(const T&) 
  [with T = double; 
  typename std::enable_if<std::is_floating_point<typename std::decay<_Tp>::type>::value, void>::type = void]’
   87 |   var_value(const T& x) : vi_(new vari_type(x, false)) {}  // NOLINT
```